Add Testing Docs for v0.18

kevinmeredith opened the pull request:
From https://github.com/http4s/http4s/pull/1699

Could you please take a look, @rossabaker, @jcranky, @jmcardon, @ChristopherDavenport? Thanks!

Closes https://github.com/http4s/http4s/issues/1504

rossabaker commented:
The test failure is not yours.

jcranky commented on docs/src/main/tut/testing.md:
```diff
@@ -0,0 +1,155 @@
+
+final case class User(name: String, age: Int)
+
+implicit val UserEncoder: Encoder[User] = deriveEncoder[User]
```
Should this top level `val` be here? It doesn't compile, does it?

jcranky replied:
Ah, it will compile as in a REPL session, but not on "normal" code. This is a bit confusing.

kevinmeredith replied:
I originally put the `Encoder[User]` in `object User`, i.e. the companion. However, if I recall correctly, I got a `tut` error due to, as I understand, the `implicit` not being in scope, i.e. due to the REPL.

Please let me know what you suggest, @jcranky, to make it clear!

jmcardon replied:
Worst case:

Put the implicit val in the companion obj, then `import User._ //note: this is necessary for the repl`.

Pretty much all you need.

jcranky replied:
@jmcardon 's suggestion would be enough. Other than that, I recall that there were some trick to make that work with repl, like the `:paste` mode, but I don't know how to express that for `tut`

kevinmeredith commented:
I moved the `Encoder[User]` per your comments, @jcranky and @jmcardon. 

Please take a look again at this PR! Thank you.

rossabaker commented on docs/src/main/tut/testing.md:
```diff
@@ -0,0 +1,157 @@
+  implicit val UserEncoder: Encoder[User] = deriveEncoder[User]
+}
+
+import User._ // Note: this is necessary for the REPL.
```
Because of https://github.com/tpolecat/tut/issues/62?

kevinmeredith replied:
That's my understanding per https://github.com/http4s/http4s/pull/1718#discussion_r176492555.

rossabaker commented on docs/src/main/tut/testing.md:
````diff
@@ -0,0 +1,157 @@
+}
+```
+
+For testing, let's define a `assert` function:
````
Do you mean `check`?

rossabaker commented on docs/src/main/tut/testing.md:
```diff
@@ -0,0 +1,157 @@
+The above documentation demonstrated how to define an HttpService[F], pass `Request`'s, and then 
+test the expected `OptionT[F, Response]`.
+
+To add unit tests in your choice Scala Testing Framework, please follow the above examples.
```
s/choice/chosen/ ?

rossabaker commented on docs/src/main/tut/testing.md:
```diff
@@ -0,0 +1,157 @@
+
+## Example
+
+Given the following service, which we can define through the following [Swagger](https://swagger.io/) contract:
```
I like the swagger use, but is it really relevant to testing?  We're not generating a client off this. :thinking:

kevinmeredith replied:
I added a Swagger doc for a few reasons:

1. habit - at work, before developing a new microservice, we draw up a contract, i.e. Swagger doc, first 
2. I thought it would be helpful to document the service with its contract before implementing it.

Please let me know if you don't think it adds any value. I'm not going to fight to keep it - just want the docs to be clear.

rossabaker replied:
I think that approach to developing services has some merit.  It just feels a bit like a distraction from the tut, to me.  I'm mildly in favor of its removal, but if other people like it here, I also won't fight it.

jcranky replied:
I'm not against keeping it, but I have to say I wouldn't read it when reading the tut. Yaml is nice but is still noisy and I only really read it via some nice UI.

kevinmeredith commented:
> Most of our unit tests call orNotFound on the service to "seal" it, and then they don't need the OptionT complexity at all. Is this worth mentioning?

Definitely. Come to think of it, that's what my colleagues and I do presently at work on our unit tests. I'll add that change, i.e. using `orNotFound` rather than deal with the `OptionT`.

kevinmeredith commented:
I believe that I've addressed your helpful comments, @rossabaker.

In my recent commits, I had moved the `Encoder[User]` onto the REPL's root-level due to an issue that I described in https://github.com/http4s/http4s/pull/1718/commits/104aff4bafeacf96292cee81be7331adf1474e32.

As I understand, the only pending item is whether to keep or scrap the Swagger Doc. Care to weigh in, please, @ChristopherDavenport or @jmcardon, in this thread https://github.com/http4s/http4s/pull/1718#discussion_r176748965?

rossabaker commented on docs/src/main/tut/testing.md:
````diff
@@ -135,11 +130,25 @@ val foundNone: UserRepo[IO] = new UserRepo[IO] {
+check(response, Status.NotFound, None)
+```
+
+Finally, let's pass a `Request` for which our service does not handle it:  
````
Diction here is a bit awkward.  "Finally, let's pass a`Request` which our service does not handle.", maybe?

kevinmeredith replied:
Yes - that's not clear at all as-is. I pushed up a change per your suggestion.

kevinmeredith commented:
Also, I tried to preview my changes via `previewSite`, yet it failed:

```
>project docs
[docs] λ previewSite
[info] Running tut.TutMain /Users/kevinmeredith/Workspace/http4s/docs/src/main/tut/testing.md /Users/kevinmeredith/Workspace/http4s/docs/target/scala-2.12/tut .*\.(md|markdown|txt|htm|html) -deprecation -encoding utf-8 -explaintypes -feature -language:existentials -language:experimental.macros -language:higherKinds -language:implicitConversions -unchecked -Xfatal-warnings -Xfuture -Yno-adapted-args -Ywarn-dead-code -Ywarn-inaccessible -Ywarn-nullary-override -Ywarn-nullary-unit -Ywarn-numeric-widen -Ywarn-value-discard -Xlint:constant -Ywarn-unused:implicits -Ywarn-unused:locals -Ywarn-unused:patvars -Ywarn-unused:privates -Ywarn-extra-implicit -Xlint:adapted-args -Xlint:by-name-right-associative -Xlint:delayedinit-select -Xlint:doc-detached -Xlint:inaccessible -Xlint:infer-any -Xlint:missing-interpolator -Xlint:nullary-override -Xlint:nullary-unit -Xlint:option-implicit -Xlint:package-object-classes -Xlint:poly-implicit-overload -Xlint:private-shadow -Xlint:stars-align -Xlint:type-parameter-shadow -Xlint:unsound-match -Ywarn-infer-any -Ypartial-unification -Xplugin:/Users/kevinmeredith/.coursier/cache/v1/https/repo1.maven.org/maven2/org/scalamacros/paradise_2.12.4/2.1.1/paradise_2.12.4-2.1.1.jar -Xplugin:/Users/kevinmeredith/.coursier/cache/v1/https/repo1.maven.org/maven2/org/spire-math/kind-projector_2.12/0.9.6/kind-projector_2.12-0.9.6.jar -target:jvm-1.8 -Xfatal-warnings
[tut] compiling: /Users/kevinmeredith/Workspace/http4s/docs/src/main/tut/testing.md
[info] Main Scala API documentation to /Users/kevinmeredith/Workspace/http4s/docs/target/scala-2.12/unidoc...
model contains 698 documentable templates
[info] Main Scala API documentation successful.
[error] java.io.IOException: Cannot run program "hugo": error=2, No such file or directory
[error] 	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
[error] 	at scala.sys.process.ProcessBuilderImpl$Simple.run(ProcessBuilderImpl.scala:71)
[error] 	at scala.sys.process.ProcessBuilderImpl$AbstractBuilder.$bang(ProcessBuilderImpl.scala:115)
[error] 	at scala.sys.process.ProcessBuilderImpl$AbstractBuilder.slurp(ProcessBuilderImpl.scala:131)
[error] 	at scala.sys.process.ProcessBuilderImpl$AbstractBuilder.$bang$bang(ProcessBuilderImpl.scala:104)
[error] 	at com.typesafe.sbt.site.hugo.HugoPlugin$.$anonfun$unsafeCheckHugoVersion$2(HugoPlugin.scala:69)
[error] 	at scala.util.Try$.apply(Try.scala:209)
[error] 	at com.typesafe.sbt.site.hugo.HugoPlugin$.unsafeCheckHugoVersion(HugoPlugin.scala:69)
[error] 	at com.typesafe.sbt.site.hugo.HugoPlugin$.$anonfun$hugoSettings$6(HugoPlugin.scala:35)
[error] 	at com.typesafe.sbt.site.hugo.HugoPlugin$.$anonfun$hugoSettings$6$adapted(HugoPlugin.scala:35)
[error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:44)
[error] 	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:39)
[error] 	at sbt.std.Transform$$anon$4.work(System.scala:66)
[error] 	at sbt.Execute.$anonfun$submit$2(Execute.scala:262)
[error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:16)
[error] 	at sbt.Execute.work(Execute.scala:271)
[error] 	at sbt.Execute.$anonfun$submit$1(Execute.scala:262)
[error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:174)
[error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:36)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[error] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[error] 	at java.lang.Thread.run(Thread.java:745)
```

I'm assuming that `hugo` is a binary that I need to install? Please let me know how to resolve this error. Thanks!

rossabaker commented on docs/src/main/tut/testing.md:
````diff
@@ -135,11 +130,25 @@ val foundNone: UserRepo[IO] = new UserRepo[IO] {
+```tut:book
+val doesNotMatter: UserRepo[IO] = new UserRepo[IO] {
+  def find(id: String): IO[Option[User]] = IO.raiseError(new RuntimeException("Should not get called!"))
+} 
````
Where is this used?

rossabaker commented:
[How to install Hugo](https://gohugo.io/getting-started/installing/).  It's just a single static binary that won't haunt you in your sleep, like a Jekyll installation. :ghost:

kevinmeredith commented:
> 

Thanks for the instructions! After running `previewSite`, I see the following:

<img width="1578" alt="screen shot 2018-03-25 at 9 39 55 am" src="https://user-images.githubusercontent.com/1967302/37875663-951771f8-3010-11e8-8c42-2a77dd316d10.png">

This doesn't look right, i.e. it's missing the other `MD` links on the right-hand side.

Perhaps I'm doing something wrong? Please help me!

rossabaker commented:
I'm not sure what's going on in that site preview.  Try running `tut` first.  I think `previewSite` might use `tutQuick`.  I wonder whether it's just not building all of them.

kevinmeredith commented:
I ripped out the Swagger Docs from the `testing.md`.

Both `tutQuick` and `previewSite` and `tut` and `previewSite` produced the same output HTML as https://github.com/http4s/http4s/pull/1718#issuecomment-375971428.

Do you think that the `previewSite` is necessary before merging this PR, @rossabaker? Thanks

rossabaker commented on docs/src/main/tut/testing.md:
```diff
@@ -6,60 +6,12 @@ title: Testing
 ## Introduction
 
-This document will show a simple Swagger Doc, its implementation as an `org.http4s.HttpService`, and then
+This document implement a simple `org.http4s.HttpService` and then
```
*implements

kevinmeredith replied:
Thanks - good 👁!
